### PR TITLE
feat: single env file

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes among others #1083 #1069 #1021

cc @iduuck 

## ✅ Checklist

- [ ] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [ ] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

Change to use a single env file as entry point. No need to import from server/client - we use a proxy to make sure the right errors are thrown if server vars are accessed on the client.

---

## Screenshots

_[Screenshots]_

For example, if you try accessing `env.DATABASE_URL` on the client:

image

### CODE REVIEW

File: cli/template/extras/src/env/with-auth.mjs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+/** @type z.infer<merged>
+ *  @ts-ignore - can't type this properly in jsdoc */
+export const env = new Proxy(parsed.data, {
+  get(target, prop) {
+    if (typeof prop !== "string") return undefined;
+    // Throw a descriptive error if a server-side env var is accessed on the client
+    // Otherwise it would just be returning `undefined` and be annoying to debug
+    if (!isServer && !prop.startsWith("NEXT_PUBLIC_"))
+      throw new Error(
+        "❌ Attempted to access serverside environment variable on the client",
```
Comment (iduuck) [inline cli/template/extras/src/env/with-auth.mjs:86]: server-side, or?

File: cli/template/extras/src/env/with-prisma.mjs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+const isServer = typeof window === "undefined";
+
+const merged = server.merge(client);
+const parsed = isServer
+  ? merged.safeParse(processEnv) // on server we can validate all env vars
+  : client.safeParse(processEnv); // on client we can only validate the ones that are exposed
+
+if (parsed.success === false) {
+  console.error(
```
Comment (iduuck) [inline cli/template/extras/src/env/with-prisma.mjs:55]: imo, we should "streamline" this (below we are using "new Error", and here we are using "console.error"..

**Comment by iduuck** [comment]
Beside my two minor comments, I am sure that we could additionally use the type overwrite of `process.env`, if we want to. Though, we don't have the server vs. client type safety this way.

**Comment by juliusmarminge** [comment]
> Beside my two minor comments, I am sure that we could additionally use the type overwrite of `process.env`, if we want to. Though, we don't have the server vs. client type safety this way.

I dont like the process.env typing since you miss stuff like default values for stuff your host provides, transforms/preprocess and other powerful primitives Zod provides

File: cli/template/extras/src/env/with-prisma.mjs (hunk 2 above)
Comment (juliusmarminge) [reply cli/template/extras/src/env/with-prisma.mjs:55]: Actually we do throw here, we just log a nice formatted message before.

CleanShot 2023-01-20 at 11 25 33

**Comment by c-ehrlich** [comment]
Which use cases do we need to test to make sure that this works as desired?
- dev
- vercel
- docker (passing in via docker-compose)
- edge?
- other?

**Comment by juliusmarminge** [comment]
> Which use cases do we need to test to make sure that this works as desired?
> 
> * dev
> * vercel
> * docker (passing in via docker-compose)
> * edge?
> * other?

Got one here: https://github.com/juliusmarminge/env-test

Tested on

- vercel edge: https://env-test-lyart.vercel.app/api/edge
- vercel lambda: https://env-test-lyart.vercel.app/api/lambda
- local dev (both edge + lambda)

Go to https://env-test-lyart.vercel.app/server and it will crash since we try accessing server var

**Comment by iduuck** [comment]
I think we need to check for the following use cases:

- middleware (see #1000)
- vercel
- docker
- edge functions
- lambda
- cloudflare workers (?)

File: cli/template/extras/src/env/with-prisma.mjs (hunk 2 above)
Comment (iduuck) [reply cli/template/extras/src/env/with-prisma.mjs:55]: I may have missed this, but perhaps also there a nice formatted message, so the redbox from Next.js is going to tell us straight which are wrong, and we do not need to go into console?

**Comment by iduuck** [comment]
One more thought from my side: In the production environment, we might not want to expose which variables are wrong (see your crashing server url):
Screenshot 2023-01-21 at 12 46 20

**Comment by juliusmarminge** [comment]
> One more thought from my side: In the production environment, we might not want to expose which variables are wrong (see your crashing server url):
> Screenshot 2023-01-21 at 12 46 20

Yea I was hoping this would error on build tbf when collecting the page data 🤔

**Comment by c-ehrlich** [comment]
This implementation is still crashing during docker build if you call next-auth stuff in GSSP (see https://github.com/t3-oss/create-t3-app/issues/992#issuecomment-1404043116)

I got a working solution, but there's probably a cleaner way of writing it: https://github.com/c-ehrlich/docker-next-auth/blob/main/src/env.mjs

**Comment by bai** [comment]
Thanks for doing this, a very welcome change ❤️

**Comment by KaiSpencer** [comment]
Webstorm infers `env` as `any` but VSCode correctly infers the type from the merged zod schema. Any ideas on this? :D

**Comment by iduuck** [comment]
What does tsc give you here in the project? Seems like WebStorm isn't using the right jsdoc types.

**Comment by c-ehrlich** [comment]
Added changes from https://github.com/t3-oss/create-t3-app/pull/1092#issuecomment-1404044831
- works in docker
- allows calling next-auth in gssp without failing in build

Tested in local dev and docker

**Review by joachimjusth — APPROVED** [review]
Ok for french translation update

File: www/src/pages/ar/usage/env-variables.md [hunk 3]
```diff
@@ -74,7 +74,7 @@ const validated = schema.parse(process.env);
 إذا أردت إستخدام الـ env vars فيمكنك إستيراد` env/client.mjs` و `env/server.mjs` في المكان الذي تريد
```
Comment (Sboonny) [inline www/src/pages/ar/usage/env-variables.md:74]:
```suggestion
إذا أردت إستخدام الـ env vars فيمكنك إستيراد` env.mjs` واستعمالهم طبيعيا. إذا إستيرادت الملف في Client وحولت استعمال قيم الserver-side، ستنذر بوجود خطأ في run-time.
```

File: cli/template/base/src/env.mjs [hunk 4]
```diff
@@ -31,45 +31,57 @@ const processEnv = {
...
+  const formatErrors = (
+    /** @type {z.ZodFormattedError<Map<string,string>,string>} */
+    errors,
+  ) =>
+    Object.entries(errors)
+      .map(([name, value]) => {
+        if (value && "_errors" in value)
+          return `${name}: ${value._errors.join(", ")}\n`;
+      })
+      .filter(Boolean);
```
Comment (juliusmarminge) [inline cli/template/base/src/env.mjs:49]: i think we can remove this, the default `ZodError.flatten()` looks good

Comment (c-ehrlich) [reply cli/template/base/src/env.mjs:49]: image
looks nice

**Comment by juliusmarminge** [comment]
Good on my side. Ship? @c-ehrlich ?

**Comment by c-ehrlich** [comment]
> Good on my side. Ship? @c-ehrlich

yea good for me also

**Review by c-ehrlich — APPROVED** [review]

**Review by Sn0wye — COMMENTED** [review]
LGTM!

**Pull request merged**